pkg/ipcache/metadata.go

```diff
@@ -0,0 +1,406 @@
+// not modifying the returned object as it's a live reference to the underlying
+// map.
+func GetIDMetadataByIP(prefix string) labels.Labels {
+	idMDMU.RLock()
+	defer idMDMU.RUnlock()
+	return IdentityMetadata[prefix]
+}
+
+// InjectLabels injects labels into the IdentityMetadata (IDMD) map. The given
+// source is the source of the caller, as inserting into the IPCache requires
+// where this updated information is coming from.
+
+// Note that as this function iterates through the IDMD, if it detects a change
+// in labels for a given prefix, then this might allocate a new identity. If a
+// prefix was previously associated with an identity, it will get deallocated,
+// so a balance is kept.
+func InjectLabels(src source.Source, updater identityUpdater, triggerer policyTriggerer) error {
+	if IdentityAllocator == nil || !IdentityAllocator.IsLocalIdentityAllocatorInitialized() {
+		return ErrLocalIdentityAllocatorUninitialized
+	}
+
+	if IPIdentityCache.k8sSyncedChecker == nil ||
+		!IPIdentityCache.k8sSyncedChecker.K8sCacheIsSynced() {
+		return errors.New("k8s cache not fully synced")
+	}
+
+	if updater == nil || triggerer == nil {
+		return errors.New("policy updater not yet initialized")
+	}
+
+	idMDMU.Lock()
+	defer idMDMU.Unlock()
+
+	var (
+		// trigger is true when we need to trigger policy recalculations.
+		trigger bool
+		// toUpsert stores IPKeyPairs to upsert into the ipcache.
+		toUpsert = make(map[string]Identity)
+		// idsToPropagate stores the identities that must be updated via the
+		// selector cache.
+		idsToPropagate = make(map[identity.NumericIdentity]labels.LabelArray)
+	)
+	for prefix, lbls := range IdentityMetadata {
+		id, isNew, err := injectLabels(prefix, lbls)
+		if err != nil {
+			return fmt.Errorf("failed to allocate new identity for IP %v: %w", prefix, err)
+		}
+
+		// Reserved IDs should always be upserted if there was a change to
+		// their labels. This is especially important for IDs such as
+		// kube-apiserver which is can be accompanied by other labels such as
+		// remote-node, host, or CIDR labels.
+		// Also, any new identity should be upserted.
+		if id.IsReserved() || isNew {
+			var hasKubeAPIServer bool
+			if lbls.Has(labels.LabelKubeAPIServer[labels.IDNameKubeAPIServer]) {
+				// Overwrite the source because any IP associated with the
+				// kube-apiserver takes the strongest precedence. This is
+				// because we need to overwrite Local if only the local node IP
+				// has been upserted into the ipcache first.
+				//
+				// Also, trigger policy recalculations to update kube-apiserver
+				// identity.
+				src = source.KubeAPIServer
+				trigger = true
+				hasKubeAPIServer = true
+			}
+
+			toUpsert[prefix] = Identity{
+				ID:     id.ID,
+				Source: src,
+			}
+			if id.IsReserved() && hasKubeAPIServer {
+				identity.AddReservedIdentityWithLabels(id.ID, lbls)
+				idsToPropagate[id.ID] = lbls.LabelArray()
+			}
```

Comment:
(I've removed the need to use `hasKubeAPIServer`)

```suggestion
			if lbls.Has(labels.LabelKubeAPIServer[labels.IDNameKubeAPIServer]) {
				// Overwrite the source because any IP associated with the
				// kube-apiserver takes the strongest precedence. This is
				// because we need to overwrite Local if only the local node IP
				// has been upserted into the ipcache first.
				//
				// Also, trigger policy recalculations to update kube-apiserver
				// identity.
				src = source.KubeAPIServer
				trigger = true
				if id.IsReserved() {
					identity.AddReservedIdentityWithLabels(id.ID, lbls)
					idsToPropagate[id.ID] = lbls.LabelArray()
				}
			}

			toUpsert[prefix] = Identity{
				ID:     id.ID,
				Source: src,
			}
```